BuildResidentialScheduleFile/measure.rb

```diff
@@ -0,0 +1,182 @@
+    arg.setDefaultValue('smooth')
+    args << arg
+
+    arg = OpenStudio::Measure::OSArgument.makeStringArgument('schedules_vacancy_period', false)
+    arg.setDisplayName('Schedules: Vacancy Period')
+    arg.setDescription('Specifies the vacancy period. Enter a date like "Dec 15 - Jan 15".')
+    args << arg
+
+    arg = OpenStudio::Measure::OSArgument.makeIntegerArgument('schedules_random_seed', false)
+    arg.setDisplayName('Schedules: Random Seed')
+    arg.setUnits('#')
+    arg.setDescription("This numeric field is the seed for the random number generator. Only applies if the schedules type is 'stochastic'.")
+    args << arg
+
+    arg = OpenStudio::Measure::OSArgument.makeStringArgument('output_csv_path', true)
+    arg.setDisplayName('Schedules: Path')
+    arg.setDescription('Absolute (or relative) path of the csv file containing user-specified occupancy schedules.')
+    args << arg
+
+    return args
+  end
+
+  # define what happens when the measure is run
+  def run(model, runner, user_arguments)
+    super(model, runner, user_arguments)
+
+    # use the built-in error checking
+    if !runner.validateUserArguments(arguments(model), user_arguments)
+      return false
+    end
+
+    # assign the user inputs to variables
+    args = get_argument_values(runner, arguments(model), user_arguments)
+    args = Hash[args.collect { |k, v| [k.to_sym, v] }]
+
+    hpxml_path = args[:hpxml_path]
+    unless (Pathname.new hpxml_path).absolute?
+      hpxml_path = File.expand_path(File.join(File.dirname(__FILE__), hpxml_path))
+    end
+    unless File.exist?(hpxml_path) && hpxml_path.downcase.end_with?('.xml')
+      fail "'#{hpxml_path}' does not exist or is not an .xml file."
+    end
+
+    hpxml = HPXML.new(hpxml_path: hpxml_path)
+
+    # Create EpwFile object
+    epw_path = hpxml.climate_and_risk_zones.weather_station_epw_filepath
+    if not File.exist? epw_path
+      epw_path = File.join(File.expand_path(File.join(File.dirname(__FILE__), '..', 'weather')), epw_path) # a filename was entered for weather_station_epw_filepath
+    end
+    if not File.exist? epw_path
+      runner.registerError("Could not find EPW file at '#{epw_path}'.")
+      return false
+    end
+    epw_file = OpenStudio::EpwFile.new(epw_path)
+
+    success = create_schedules(runner, hpxml, model, epw_file, args)
+    return false if not success
+
+    # modify the hpxml with the schedules path
+    hpxml.header.schedules_filepath = args[:output_csv_path]
+    hpxml_doc = hpxml.to_oga()
+    XMLHelper.write_file(hpxml_doc, hpxml_path)
+    runner.registerInfo("Wrote file: #{hpxml_path}")
```

Comment:
We shouldn't use the HPXML class to update the HPXML file, as the class is not comprehensive and does not cover every possible HPXML element. If someone had created their own custom HPXML file (i.e., not through the HPXML class) and then you read in and write the file using the HPXML class, information will be lost. Rather, you should use the XMLHelper or Oga methods directly to insert/update the schedules path.

(It is fine to read in the HPXML file using the HPXML class in order to retrieve the state code, etc. We just can't use it to update/write the file.)

Reply:
@shorowit You mentioned "[we] might need to add an argument to the write_file method to bypass a lot of the stuff it's doing". Can you implement that here?